src/zap/zapreadytorun.cpp

```diff
@@ -332,6 +332,74 @@ void ZapImage::OutputDebugInfoForReadyToRun()
     GetReadyToRunHeader()->RegisterSection(READYTORUN_SECTION_DEBUG_INFO, pBlob);
 }
 
+void ZapImage::OutputTypesTableForReadyToRun(IMDInternalImport * pMDImport)
+{
+    NativeWriter writer;
+    VertexHashtable typesHashtable;
+
+    NativeSection * pSection = writer.NewSection();
+    pSection->Place(&typesHashtable);
+
+    // Save all typedef entries into hashtable
+    {
+        // Note on duplicate types with same name: there is not need to perform that check when building
+        // the hashtable. If such types were encountered, the R2R compilation would fail before reaching here.
+
+        for (int i = 0; i < 2; i++)
+        {
+            HENUMInternalHolder hEnum(pMDImport);
+            i == 0 ? hEnum.EnumAllInit(mdtTypeDef) : hEnum.EnumInit(mdtExportedType, mdTokenNil);
+
+            mdToken mdTypeToken;
+            while (pMDImport->EnumNext(&hEnum, &mdTypeToken))
+            {
+                DWORD dwHash = 0;
+                BOOL hasEncloser;
+                mdTypeDef mdCurrentToken = mdTypeToken;
+
+                do
+                {
+                    LPCUTF8        pszName;
+                    LPCUTF8        pszNameSpace;
+                    if (i == 0)
+                    {
+                        if (FAILED(pMDImport->GetNameOfTypeDef(mdCurrentToken, &pszName, &pszNameSpace)))
+                            ThrowHR(COR_E_BADIMAGEFORMAT);
+
+                        hasEncloser = SUCCEEDED(pMDImport->GetNestedClassProps(mdCurrentToken, &mdCurrentToken));
+                    }
+                    else
+                    {
+                        if (FAILED(pMDImport->GetExportedTypeProps(mdCurrentToken, &pszNameSpace, &pszName, &mdCurrentToken, NULL, NULL)))
+                            ThrowHR(COR_E_BADIMAGEFORMAT);
+
+                        hasEncloser = (TypeFromToken(mdCurrentToken) == mdtExportedType);
+                    }
+
+                    if (pszNameSpace != NULL)
+                        dwHash = ((dwHash << 5) + dwHash) ^ HashStringA(pszNameSpace);
+                    dwHash = ((dwHash << 5) + dwHash) ^ HashStringA(pszName);
+
```

Comment:
It would be nice to use the same type hashcode algorithm as everywhere (TypeHashingAlgorithms.cs).